File: 2141-maximum-running-time-of-n-computers/2141-maximum-running-time-of-n-computers.cpp
typedef long long ll;

class Solution {
public:
    long long maxRunTime(int n, vector<int>& batteries) {
        ll left = 1, right = accumulate(batteries.begin(), batteries.end(), 0ll) / n;
        
        while (left < right){
            ll target = right - ((right - left) >> 1), extra = 0;
            
            for (int power : batteries) extra += min((ll)power, target);
            
            if (extra / n >= target) left = target;
            else right = target - 1;
        }
        
        return left;
    }
};

File: 2141-maximum-running-time-of-n-computers/2141-maximum-running-time-of-n-computers.py
class Solution:
    def maxRunTime(self, n: int, batteries: List[int]) -> int:
        left, right = 1, sum(batteries) // n
        
        while left < right:
            target = right - ((right - left)>> 1)
            extra = 0
            
            # accumulate sum of min batteries
            for power in batteries:
                extra += min(power, target)
            
            if extra // n >= target:
                left = target
            else:
                right = target - 1
        
        return left